liquidator/test/Liquidator.js

```diff
@@ -430,6 +435,216 @@ contract("Liquidator.js", function(accounts) {
     assert.equal(spy.callCount, 2); // 1 new log level event due to the successful execution.
   });
 
+  describe("Partial liquidations", function() {
+    it("amount-to-liquidate > min-sponsor-tokens", async function() {
+      // We'll attempt to liquidate 6 tokens. The minimum sponsor position is 5. There are 3 different scenarios
+      // we should test for, each of which we'll attempt to liquidate in one call of `queryAndLiquidate`.
+      const amountToLiquidate = toWei("6");
+
+      // 1. (tokens-outstanding - amount-to-liquidate) > min-sponsor-tokens, and amount-to-liquidate < tokens-outstanding
+      //     - The bot will be able to liquidate its desired amount, leaving the position above the minimum token threshold.
+      //     - Example: (12 - 6) > 5, new position will have 6 tokens remaining.
+      await emp.create({ rawValue: toWei("100") }, { rawValue: toWei("12") }, { from: sponsor1 });
+      // 2. (tokens-outstanding - amount-to-liquidate) <= min-sponsor-tokens, and amount-to-liquidate < tokens-outstanding
+      //     - The bot will NOT be able to liquidate its desired amount. It will liquidate a reduced amount and
+      //       reduce the position exactly to the minimum.
+      //     - Example: (8 - 6) <= 5, so instead the bot will liquidate (8 - 5) = 3 tokens to leave the position with (8 - 3) = 5 tokens remaining.
+      await emp.create({ rawValue: toWei("100") }, { rawValue: toWei("8") }, { from: sponsor2 });
+      // 3. amount-to-liquidate > tokens-outstanding
+      //     - The bot will liquidate the full position.
+      //     - Example: 6 > 5, so the bot will liquidate 5 tokens.
+      await emp.create({ rawValue: toWei("100") }, { rawValue: toWei("5") }, { from: sponsor3 });
+
+      // liquidatorBot creates a position with enough tokens to liquidate all positions.
+      await emp.create({ rawValue: toWei("10000") }, { rawValue: toWei("50") }, { from: liquidatorBot });
+
+      // Next, assume the price feed given to the liquidator has moved such that the sponsors
+      // are all now undercollateralized. The liquidator bot should correctly identify this and liquidate the positions.
+      // A price of 25 USD per token will make all positions undercollateralized.
+      // Numerically debt * price * coltReq > debt for collateralized position.
+      // Sponsor1: 12 * 25 * 1.2 > 100
+      // Sponsor2: 8 * 25 * 1.2 > 100
+      // Sponsor3: 5 * 25 * 1.2 > 100
+      priceFeedMock.setCurrentPrice(toBN(toWei("25")));
+
+      await liquidator.queryAndLiquidate(amountToLiquidate);
+
+      // Check logs are emitted correctly. Partial liquidations should emit an "error"-level alert before a normal liquidation "info"-level alert.
+      assert.equal(spy.callCount, 5); // 3 info + 2 error level events should be sent at the conclusion of the 3 liquidations including 2 partials.
+      assert.equal(spyLogLevel(spy, 4), "info");
```

Comment:
I broke down log level unit tests more clearly here to show that partial liquidations send "error"-level alerts

Reply:
+1